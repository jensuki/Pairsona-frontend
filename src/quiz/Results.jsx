import { useState, useEffect, useContext } from 'react';
import Api from '../helpers/api';
import UserContext from '../context/UserContext';
import Loading from '../components/Loading';
import '../styles/Results.css';

const Results = () => {
    const { currUser } = useContext(UserContext);
    const [result, setResult] = useState(null);
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(true); // use local state

    useEffect(() => {
        const fetchResult = async () => {
            try {
                const data = await Api.getMbtiDetails(currUser.username);
                setResult(data);
            } catch (err) {
                setError(err.message);
            } finally {
                setLoading(false);
            }
        };

        fetchResult();
    }, [currUser]);

    if (loading) return <Loading />
    if (error) return <div className="error">{error}</div>;
    if (!result) return <div>No results available. Please take the quiz.</div>;

    return (
        <div className="ResultsPage">
            <div className="results-container">
                {/* quiz results */}
                <h2>Your result:</h2>
                <h3><strong>{result.type}</strong></h3>
                <h4>{result.title}</h4>
                <p>Percentage of Population: {result.percentage}</p>
                <p className="result-desc">{result.description}</p>
                <a href={result.site} target="_blank" rel="noreferrer">
                    Learn more about {result.type}
                </a>
            </div>
        </div>
    );
};

export default Results;
