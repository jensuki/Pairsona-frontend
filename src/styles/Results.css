.ResultsPage {
    max-width: var(--max-width-md);
}

.ResultsPage h2 {
    font-size: 1.2rem;
    margin: 0 0 20px 0;
}

.ResultsPage h3 {
    font-size: 1.4rem;
}

.ResultsPage h4 {
    font-size: 1.4rem;
    font-style: italic;
    margin-bottom: 0;
}

.result-desc {
    font-size: 1.4rem;
}

.ResultsPage strong {
    background: rgb(115, 102, 104);
    padding: .3rem .7rem;
    border-radius: var(--border-radius-md);
    font-size: 1.5em;
    color: white;
}