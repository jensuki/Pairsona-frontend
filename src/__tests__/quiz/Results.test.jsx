import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import Results from '../../quiz/Results';
import Api from '../../helpers/api';
import { describe, test, expect, vi, beforeEach } from 'vitest';
import { demoUser } from '../../test_common/mockData';

// mock api funcs
vi.mock('../../helpers/api', () => ({
    default: {
        getMbtiDetails: vi.fn()
    }
}))

// helper
const renderResults = () => {
    return render(
        <MemoryRouter>
            <UserProvider currUser={demoUser}>
                <Results />
            </UserProvider>
        </MemoryRouter>
    )
}

describe('Results component', () => {
    beforeEach(() => {
        vi.clearAllMocks()
    })

    test('renders without crashing', async () => {
        renderResults();

        // expect spinner
        await waitFor(() => {
            expect(screen.getByRole('status')).toBeInTheDocument();
        })
    })

    test('matches snpashot initially', async () => {
        const { asFragment } = renderResults();

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        })
    })

    test('matches snapshot with results', async () => {
        // mock successful api response
        Api.getMbtiDetails.mockResolvedValueOnce({
            type: "INFJ",
            title: "The Advocate",
            percentage: "1.5%",
            description: "Advocates are quiet visionaries, often serving as inspiring and tireless idealists.",
            site: "https://www.16personalities.com/infj-personality"
        })

        const { asFragment } = renderResults();

        // wait for results to appear
        await waitFor(() => {
            expect(screen.getByText(/Your Result:/i)).toBeInTheDocument();
        })

        expect(asFragment()).toMatchSnapshot();
    })
    test('displays the mbti type along with results on the page', async () => {
        // mock successful api response
        Api.getMbtiDetails.mockResolvedValueOnce({
            type: "INFJ",
            title: "The Advocate",
            percentage: "1.5%",
            description: "Advocates are quiet visionaries, often serving as inspiring and tireless idealists.",
            site: "https://www.16personalities.com/infj-personality"
        });

        renderResults();

        // wait for results to load
        await waitFor(() => {
            expect(screen.getByText(/Your result:/i)).toBeInTheDocument();

            // verify "INFJ" h3 element is displayed
            const mbtiType = screen.getByRole('heading', { level: 3 });
            expect(mbtiType).toHaveTextContent('INFJ');
        })
    });

    test('displays error message on API failure', async () => {
        // mock error msg
        Api.getMbtiDetails.mockRejectedValueOnce(new Error('Failed to fetch results'));

        renderResults();

        await waitFor(() => {
            expect(screen.getByText(/failed to fetch results/i)).toBeInTheDocument();
        });

        // Ensure loading spinner is removed
        expect(screen.queryByRole('status')).not.toBeInTheDocument();
    });
})