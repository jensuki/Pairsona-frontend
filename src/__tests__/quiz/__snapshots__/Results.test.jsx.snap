// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Results component > matches snapshot with results 1`] = `
<DocumentFragment>
  <div
    class="ResultsPage"
  >
    <div
      class="results-container"
    >
      <h2>
        Your result:
      </h2>
      <h3>
        <strong>
          INFJ
        </strong>
      </h3>
      <h4>
        The Advocate
      </h4>
      <p>
        Percentage of Population: 1.5%
      </p>
      <p
        class="result-desc"
      >
        Advocates are quiet visionaries, often serving as inspiring and tireless idealists.
      </p>
      <a
        href="https://www.16personalities.com/infj-personality"
        rel="noreferrer"
        target="_blank"
      >
        Learn more about INFJ
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Results component > matches snpashot initially 1`] = `
<DocumentFragment>
  <div
    class="loading-container"
    role="status"
  >
    <i
      class="fas fa-spinner fa-spin fa-3x"
    />
  </div>
</DocumentFragment>
`;
